# sraX User Guide: real data examples

## 1) User-provided data for AMR DB compilation
### Public ARG sequences repositories:
A recently published work [[1]](https://doi.org/10.1093/bioinformatics/bty987) describes a toolkit ([[ARGDIT]](https://github.com/phglab/ARGDIT)) for creating curated AMR DBs. The authors provided already integrated AMR DBs as examples, and this valuable information is going to be employed for demonstrating the practicality and utility of **sraX** for resistome profiling.



### User's own ARG sequences:

## 2) Genome data acquisition from already analyzed collections  
### Data-set 1:
The authors look at the phenotype and genotype correlation in _Salmonella enterica_ with different antibiotic resistance patterns.

https://www.ncbi.nlm.nih.gov/assembly?LinkName=bioproject_assembly_all&from_uid=242614#

### Data-set 2: 

### Data-set 3: 

## References.
[1] Jimmy Ka Ho Chiu, Rick Twee-Hee Ong; ARGDIT: a validation and integration toolkit for Antimicrobial Resistance Gene Databases, _Bioinformatics_, , bty987, 

[2] 

[3] 

[4]

[5]
